### PR TITLE
ADD many missing clojure.core macros to font-lock declaration

### PR DESCRIPTION
**Pull request by boxxxie**
#622 

font locks added

```
"gen-class"
"while"
"import"
"pvalues"
"bound-fn"
"vswap!"
"with-loading-context"
"delay"
"gen-interface"
"with-bindings"
"deftype"
"with-precision"
"lazy-seq"
"let"
"defstruct"
"areduce"
"definline"
"future"
"fn"
"definterface"
"some->>"
"amap"
"extend-type"
"defmethod"
"time"
"memfn"
"extend-protocol"
"reify"
"defonce"
"defn-"
"defprotocol"
"sync"
"assert"
"proxy-super"
"loop"
"with-out-str"
"with-in-str"
"some->"
"proxy"
"locking"
"defmulti"
"io!"
"lazy-cat"
"comment"
"defrecord"
"refer-clojure"
```

### CODE REVIEW

**Comment by boxxxie** [comment]
CI fail is unrelated to my changes, i dono what to do

File: clojure-mode.el [hunk 1]
```diff
@@ -841,15 +841,20 @@ any number of matches of `clojure--sym-forbidden-rest-chars'."))
...
-          '("letfn" "case" "cond" "cond->" "cond->>" "condp"
-            "for" "when" "when-not" "when-let" "when-first" "when-some"
-            "if-let" "if-not" "if-some"
-            ".." "->" "->>" "as->" "doto" "and" "or"
-            "dosync" "doseq" "dotimes" "dorun" "doall"
-            "ns" "in-ns"
-            "with-open" "with-local-vars" "binding"
-            "with-redefs" "with-redefs-fn"
-            "declare") t)
+          '("->" "->>" ".." "amap" "and" "areduce" "as->" "assert" "binding"
```
Comment (vemv) [inline clojure-mode.el:844]: While we're here we could leave use one line per item, that way it should be easier to tell whether something's there

(agreed, @bbatsov ?)

Comment (boxxxie) [reply clojure-mode.el:844]: i prefer that @vemv, i just tried to make it look similar to how it previously looked.
i guess if @bbatsov says yes, i'll update the commit. or should i just update it and assume it'll be ok?

File: clojure-mode.el [hunk 2]
```diff
@@ -841,15 +841,90 @@ any number of matches of `clojure--sym-forbidden-rest-chars'."))
...
+            "binding"
+            "bound-fn"
+            "case"
+            "comment"
+            "cond"
+            "cond->"
+            "cond->>"
+            "condp"
+            "declare"
+            "definline"
```
Comment (bbatsov) [inline clojure-mode.el:862]: You can remove all the `def`something macros as they are handled in a special manner above. `fn` is handle there as well.

File: clojure-mode.el (hunk 1 above)
Comment (bbatsov) [reply clojure-mode.el]: I'm fine with with a name per line, although that'd makes things a bit inconsistent with the rest of the font-locking code.

**Comment by bbatsov** [comment]
Don't forget to add a changelog entry about this as well.

**Comment by bbatsov** [comment]
One more thing - if some of those are duplicated in `clojure-mode-extra-font-locking` they should be removed from there - e.g. `some->`.

**Comment by bbatsov** [comment]
Seems good to me. Thanks!

**Pull request merged**